<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/people/plewis/cpp/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="/people/plewis/cpp/">
    <link rel="stylesheet" type="text/css" href="/people/plewis/cpp/assets/css/main.css" />
    <title>Strom Phylogenetics C++ Tutorial: Calculating the likelihood</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-expand-lg navbar-light bg-light">
  <div class="container-fluid">
      <a class="navbar-brand" href="/people/plewis/cpp/index.html">
        <img class="navbar-logo" src="/people/plewis/cpp/assets/img/strom-logo.png" alt="Strom Tutorial Home" />
      </a>
      <button class="navbar-toggler" type="button" data-toggle="collapse" data-target=".navbar-collapse" aria-controls="#navbarNav" aria-expanded="false" aria-label="Toggle navigation">
        <span class="navbar-toggler-icon"></span>
      </button>
      <div class="collapse navbar-collapse" id="navbarNav">
        <ul class="navbar-nav mr-auto">   <!-- navbar-nav mr-auto -->
          <li class="nav-item"> 
            <a href="/people/plewis/cpp/">Home</a>
          </li>
          
          <li class="nav-item">
            <a href="/people/plewis/cpp/win/steps/">Step-by-step instructions</a>
          </li>
          
          
        </ul>
      </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">10.0 Calculating the likelihood</h1>
	<h3 class="subtitle"></h3>
    <h3 class="subtitle">(Win version)</h3>
</div>

<div class="titlebar">
	<h3 class="subsection"><a href="/people/plewis/cpp//win//steps/step-09/00-reading-and-storing-data.html">&lt;&nbsp;9.0</a> | 10.0 | <a href="/people/plewis/cpp//win//steps/step-10/01-the-likelihood-class.html">10.1&nbsp;&gt;</a></h3>
</div>

<p>Now that we can load sequence data and a tree into memory, it is time to calculate the likelihood of the tree, which is the probability of the observed data given that tree and a model of evolution.</p>

<p><a href="https://github.com/beagle-dev/beagle-lib">BeagleLib</a> (Ayres et al. 2012, 2019) is a library that provides the ability to compute phylogenetic likelihoods for any model that can be specified by an instantaneous rate matrix of a continuous-time, finite-state Markov chain. Furthermore, it allows likelihoods to be computed using Graphics Processing Units (GPUs) if they are available, which can substantially speed up likelihood evaluation for models involving many states (such as amino acid or codon models).</p>

<p>In this step of the tutorial you will use BeagleLib to compute the likelihood of a tree read from a file using the simplest evolutionary model, the Jukes-Cantor (1969) model with equal rates across sites (JC69). In later steps you will increase the complexity of the models used, ultimately being able to process both GTR+I+G models and a basic codon model (and link parameters of these models across data subsets in arbitrary ways).</p>

<div>
<table class="table table-striped" style="width:100%">
<tr class="header">
<td width="15%"><b>Step</b></td>
<td width="25%"><b>Title</b></td>
<td width="50%"><b>Description</b></td>
</tr>

		<tr class="part" id="10-1">
		<td>Step 10.1</td>
		<td><a href="/people/plewis/cpp/win/steps/step-10/01-the-likelihood-class.html">The Likelihood class</a></td>
		<td><p>Construct a Likelihood object that uses BeagleLib to calculate the log-likelihood of a tree under the simplest substitution model (the Jukes-Cantor, or JC69, model).</p>
</td>
		</tr>
	
		<tr class="part" id="10-2">
		<td>Step 10.2</td>
		<td><a href="/people/plewis/cpp/win/steps/step-10/02-test-likelihood.html">Test Likelihood</a></td>
		<td><p>Test the Likelihood class by reading and storing a tree as well as data from nexus-formatted files.</p>
</td>
		</tr>
	
</table>
</div>

<h2 id="literature-cited">Literature Cited</h2>

<p>DL Ayres, A Darling, DJ Zwickl, P Beerli, MT Holder, PO Lewis, JP Huelsenbeck, F Ronquist, DL Swofford, MP Cummings, A Rambaut, MA Suchard. 2012. BEAGLE: An application programming interface and high-performance computing library for statistical phylogenetics. Systematic Biology 61:170â€“173. <a href="https://doi.org/10.1093/sysbio/syr100">DOI: 10.1093/sysbio/syr100</a></p>

<p>DL Ayres, MP Cummings, G Baele, AE Darling, PO Lewis, DL Swofford, JP Huelsenbeck, P Lemey, A Rambaut, and MA Suchard. 2019. BEAGLE 3: improved performance, scaling, and usability for a high-performance computing library for statistical phylogenetics. Systematic Biology syz020, published April 23, 2019. <a href="https://doi.org/10.1093/sysbio/syz020">DOI: 10.1093/sysbio/syz020</a></p>

<p>TH Jukes and CR Cantor. 1969. Evolution of protein molecules. Pages 21-132 in HN Munro (ed.), Mammalian Protein Metabolism. Academic Press, New York.</p>

<div class="titlebar">
	<h3 class="subsection"><a href="/people/plewis/cpp//win//steps/step-09/00-reading-and-storing-data.html">&lt;&nbsp;9.0</a> | 10.0 | <a href="/people/plewis/cpp//win//steps/step-10/01-the-likelihood-class.html">10.1&nbsp;&gt;</a></h3>
</div>


      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/stromtutorial">GitHub</a> | <a href="/people/plewis/cpp/win/acknowledgements/">Acknowledgements</a> | <a href="/people/plewis/cpp/win/license/">License</a> | <a href="/people/plewis/cpp/win/citation/">Citation</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/people/plewis/cpp/assets/js/jquery.min.js"></script>
    <script src="/people/plewis/cpp/assets/js/highlight.js"></script>
    <script src="/people/plewis/cpp/assets/js/bootstrap.bundle.js"></script>
  </body>
</html>
